### ISSUE TITLE
Running Dask Dataframe's agg(list) on a few GBs file causes asyncio.exceptions.TimeoutError and distributed.worker - WARNINGs, or gets SIGKILL`d by OS

### ISSUE THREAD

**Issue by JoeHFB** [dask/dask#7239]
Hi Dask team,

First off, thanks for the wonderful work you guys do.  Newbie here, maybe I missed something when I was reading the docs, but I didn't see an obvious answer, so hoping you guys can help me with this issue.

**Background**:
I have a business request to ingest csv files of decently large size (1 < X < 10GBs), do transformations, data validations, ... before submitting the data to our web services.  I'm running simple POCs to see if Dask can fit here.

One of the transformations is to reduce the data by grouping by `user_id` column and getting a collection of a few other columns, so I was trying to do a simple POC of group by `user_id` and aggregate a list of `order_id` column.

**What happened**:
TLDR:
- If I run with a memory limit, `asyncio.exceptions.TimeoutError` with a whole bunch of `distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?` 
- If I run without a memory limit, 💥 - the process gets SIGKILL'd by the OS

- I tried this against csv files of 1GB, 2GB, and 5GB.
- I tried to do a simple `dataframe.groupby('user_id')['order'].agg(list)` after reading the following:
1. https://github.com/dask/dask/pull/5938
2. https://github.com/dask/dask/issues/6602
3. https://github.com/dask/dask/pull/6655
4. https://docs.dask.org/en/latest/dataframe-groupby.html#aggregate

##### Results: 
I'm doing this on a Linux Mint 20 machine with 9GB of RAM and 4 cores.

Memory limit | 1GB CSV | 2GB CSV | 5GB CSV
--------------- | ---------- | ---------- | ----------
None | took 1.5 min, used 8.9GB of ram | 💥  `Command terminated by signal 9` after a min | Didn't even try
6GB | took 7 min, used 3.3GB of ram | 💥 Got the timeout issue | Didn't even try

Since the timeout issue stacktrace, and the warnings is quite big, I've attached it to the bottom of the page.

**What you expected to happen**:

Erm, for the files to run successfully?  I was hoping to run Dask in AWS lambdas (max 10GB ram - pretty close to my machine), but obviously I'd like to get this issue resolved.

**Minimal Complete Verifiable Example**:

#### With memory limit
```python
#!/usr/bin/env python3
from dask.distributed import Client
import dask as dk
import dask.dataframe as ddf
import itertools as it
import sys

client = Client(n_workers=1, threads_per_worker=4, processes=False, memory_limit='6GB')
df = ddf.read_csv('1gb.csv', blocksize=6400000, delimiter='|')
df = df.set_index('user_id')

# https://docs.dask.org/en/latest/dataframe-groupby.html#aggregate
# https://github.com/dask/dask/pull/6655
df = df.groupby('user_id')['order_id'].agg(list)
print(df.compute())
```

#### Without memory limit
Same thing as above, I just commented out the line `client = Client(n_workers=1, threads_per_worker=4, processes=False, memory_limit='6GB')`

**Anything else we need to know?**:

Here is the STDOUT:

#### Killed without memory limit
```
$ /usr/bin/time -v ./daskTest.py
Command terminated by signal 9
	Command being timed: "./daskTest.py"
	User time (seconds): 128.36
	System time (seconds): 37.67
	Percent of CPU this job got: 111%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 2:28.38
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 9001160
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 13746
	Minor (reclaiming a frame) page faults: 9912089
	Voluntary context switches: 290751
	Involuntary context switches: 20472
	Swaps: 0
	File system inputs: 5870800
	File system outputs: 3574336
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0

```

#### The timeout error + warnings
```
distributed.worker - WARNING - Worker is at 83% memory usage. Pausing worker.  Process memory: 4.98 GB -- Worker memory limit: 6.00 GB
distributed.worker - WARNING - Worker is at 55% memory usage. Resuming worker. Process memory: 3.34 GB -- Worker memory limit: 6.00 GB
distributed.worker - WARNING - Worker is at 80% memory usage. Pausing worker.  Process memory: 4.85 GB -- Worker memory limit: 6.00 GB
distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.38 GB -- Worker memory limit: 6.00 GB
distributed.worker - WARNING - Worker is at 55% memory usage. Resuming worker. Process memory: 3.34 GB -- Worker memory limit: 6.00 GB
distributed.comm.inproc - WARNING - Closing dangling queue in <InProc  local=inproc://10.0.2.15/2711458/1 remote=inproc://10.0.2.15/2711458/9>
distributed.worker - WARNING - Worker is at 80% memory usage. Pausing worker.  Process memory: 4.85 GB -- Worker memory limit: 6.00 GB
distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 4.85 GB -- Worker memory limit: 6.00 GB
distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 4.85 GB -- Worker memory limit: 6.00 GB
distributed.worker - WARNING - Worker is at 55% memory usage. Resuming worker. Process memory: 3.34 GB -- Worker memory limit: 6.00 GB
tornado.application - ERROR - Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOLoop object at 0x7ff43fb22c40>>, <Task finished name='Task-40404' coro=<SpecCluster._correct_state_internal() done, defined at /usr/local/lib/python3.8/dist-packages/distributed/deploy/spec.py:325> exception=OSError('Timed out trying to connect to inproc://10.0.2.15/2711458/1 after 10 s')>)
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/distributed/comm/core.py", line 286, in connect
    comm = await asyncio.wait_for(
  File "/usr/lib/python3.8/asyncio/tasks.py", line 490, in wait_for
    raise exceptions.TimeoutError()
asyncio.exceptions.TimeoutError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/tornado/ioloop.py", line 741, in _run_callback
    ret = callback()
  File "/usr/local/lib/python3.8/dist-packages/tornado/ioloop.py", line 765, in _discard_future_result
    future.result()
  File "/usr/local/lib/python3.8/dist-packages/distributed/deploy/spec.py", line 406, in _close
    await self._correct_state()
  File "/usr/local/lib/python3.8/dist-packages/distributed/deploy/spec.py", line 333, in _correct_state_internal
    await self.scheduler_comm.retire_workers(workers=list(to_close))
  File "/usr/local/lib/python3.8/dist-packages/distributed/core.py", line 788, in send_recv_from_rpc
    comm = await self.live_comm()
  File "/usr/local/lib/python3.8/dist-packages/distributed/core.py", line 746, in live_comm
    comm = await connect(
  File "/usr/local/lib/python3.8/dist-packages/distributed/comm/core.py", line 308, in connect
    raise IOError(
OSError: Timed out trying to connect to inproc://10.0.2.15/2711458/1 after 10 s
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/distributed/comm/core.py", line 286, in connect
    comm = await asyncio.wait_for(
  File "/usr/lib/python3.8/asyncio/tasks.py", line 490, in wait_for
    raise exceptions.TimeoutError()
asyncio.exceptions.TimeoutError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/distributed/deploy/spec.py", line 652, in close_clusters
    cluster.close(timeout=10)
  File "/usr/local/lib/python3.8/dist-packages/distributed/deploy/cluster.py", line 110, in close
    return self.sync(self._close, callback_timeout=timeout)
  File "/usr/local/lib/python3.8/dist-packages/distributed/deploy/cluster.py", line 189, in sync
    return sync(self.loop, func, *args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/distributed/utils.py", line 340, in sync
    raise exc.with_traceback(tb)
  File "/usr/local/lib/python3.8/dist-packages/distributed/utils.py", line 324, in f
    result[0] = yield future
  File "/usr/local/lib/python3.8/dist-packages/tornado/gen.py", line 762, in run
    value = future.result()
  File "/usr/local/lib/python3.8/dist-packages/tornado/ioloop.py", line 741, in _run_callback
    ret = callback()
  File "/usr/local/lib/python3.8/dist-packages/tornado/ioloop.py", line 765, in _discard_future_result
    future.result()
  File "/usr/local/lib/python3.8/dist-packages/distributed/deploy/spec.py", line 406, in _close
    await self._correct_state()
  File "/usr/local/lib/python3.8/dist-packages/distributed/deploy/spec.py", line 333, in _correct_state_internal
    await self.scheduler_comm.retire_workers(workers=list(to_close))
  File "/usr/local/lib/python3.8/dist-packages/distributed/core.py", line 788, in send_recv_from_rpc
    comm = await self.live_comm()
  File "/usr/local/lib/python3.8/dist-packages/distributed/core.py", line 746, in live_comm
    comm = await connect(
  File "/usr/local/lib/python3.8/dist-packages/distributed/comm/core.py", line 308, in connect
    raise IOError(
OSError: Timed out trying to connect to inproc://10.0.2.15/2711458/1 after 10 s
^CError in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/distributed/client.py", line 4716, in _close_global_client
    c.close(timeout=3)
  File "/usr/local/lib/python3.8/dist-packages/distributed/client.py", line 1429, in close
    f = self.cluster.close()
  File "/usr/local/lib/python3.8/dist-packages/distributed/deploy/cluster.py", line 110, in close
    return self.sync(self._close, callback_timeout=timeout)
  File "/usr/local/lib/python3.8/dist-packages/distributed/deploy/cluster.py", line 189, in sync
    return sync(self.loop, func, *args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/distributed/utils.py", line 337, in sync
    e.wait(10)
  File "/usr/lib/python3.8/threading.py", line 558, in wait
    signaled = self._cond.wait(timeout)
  File "/usr/lib/python3.8/threading.py", line 306, in wait
    gotit = waiter.acquire(True, timeout)
KeyboardInterrupt
^C
```

**Environment**:

- Dask version:
```
$ pip3 show dask
Name: dask
Version: 2021.2.0
Summary: Parallel PyData with Task Scheduling
Home-page: https://github.com/dask/dask/
Author: None
Author-email: None
License: BSD
Location: /usr/local/lib/python3.8/dist-packages
Requires: pyyaml
Required-by: distributed
```
- Python version:
```
$ python3 -V
Python 3.8.5
```
- Operating System:
Linux Mint 20 running on virtualbox with 9GB of RAM
```
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Linuxmint
Description:	Linux Mint 20
Release:	20
Codename:	ulyana
$ uname -a
Linux joe-VirtualBox 5.4.0-26-generic #30-Ubuntu SMP Mon Apr 20 16:58:30 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

- Install method (conda, pip, source):
`pip`

**Comment by f-istvan** [comment]
What is the solution for this? I have the same issue.
